src/main/java/hudson/tasks/junit/TestResult.java

```diff
@@ -712,6 +798,16 @@ public void freeze(TestResultAction parent) {
             pr.freeze();
     }
 
+    private void addSuiteByRunAndNode(SuiteResult s) {
+        if (suitesByRunAndNode.get(s.getRunId()) == null) {
```

Comment:
This block is rather confusing to follow, because you're doing successive gets and puts, please could you consider restructuring with nesting of conditionals so it's clear where the flow of execution will go?